contracts/WardenHandler.sol

```diff
@@ -0,0 +1,311 @@
+
+    return abi.encodePacked(uint32(1), gmpPayload);
+  }
+
+  /// @notice Encode stake response
+  /// @param stakeId Stake identifier
+  /// @param stakeResult Stake result
+  function _createStakeResponse(uint64 stakeId, StakeResult memory stakeResult) private pure returns (bytes memory) {
+    return
+      _createResponse(
+        abi.encodePacked(
+          STAKE_ACTION_TYPE,
+          stakeResult.status,
+          stakeId,
+          stakeResult.reinitUnstakeId,
+          stakeResult.lpAmount
+        )
+      );
+  }
+
+  /// @notice Encode unstake response
+  /// @param status status of unstake
+  /// @param unstakeId Unstake identifier
+  /// @param reinitUnstakeId Reinited unstake identifier
+  function _createUnstakeResponse(
+    uint8 status,
+    uint64 unstakeId,
+    uint64 reinitUnstakeId
+  ) private pure returns (bytes memory) {
+    return _createResponse(abi.encodePacked(UUNSTAKE_ACTION_TYPE, status, unstakeId, reinitUnstakeId));
+  }
+
+  /// @notice Encode reinit response
+  /// @param reinitUnstakeId Reinited unstake identifier
+  function _createReinitResponse(uint64 reinitUnstakeId) private pure returns (bytes memory) {
+    return _createResponse(abi.encodePacked(REINIT_ACTION_TYPE, reinitUnstakeId));
+  }
+
+  ///@notice Handle stake request, should be implemented in Yield contract
+  ///@param stakeId Stake identifier
+  ///@param tokenAddress Address of the token
+  ///@param amountToStake Amount of tokens to stake
+  /// @return Stake reesult
+  function _handleStakeRequest(
+    uint64 stakeId,
+    address tokenAddress,
+    uint256 amountToStake
+  ) internal virtual returns (StakeResult memory);
+
+  /// @notice Handle unstake request, should be implemented in Yield contract
+  /// @param unstakeId Unstake identifier
+  /// @param lpAmount Amount of lp token to unstake
+  /// @return ustake result
+  function _handleUnstakeRequest(uint64 unstakeId, uint128 lpAmount) internal virtual returns (UnstakeResult memory);
+
+  ///@notice Handle reinit request, should be implemented in Yield contract
+  /// @return reinit result
+  function _handleReinitRequest() internal virtual returns (ReinitResult memory);
+
+  ///@notice Axelar relayer calls the function when accept unstake request from Warden
+  function execute(
+    bytes32 commandId,
+    string calldata sourceChain,
+    string calldata sourceAddress,
+    bytes calldata payload
+  ) external {
+    WardenHandlerData storage $ = _getWardenHandlerData();
+
+    string memory wardenChain = $.wardenChain;
+    if (!Strings.equal(wardenChain, sourceChain)) revert InvalidSourceChain();
+
+    string memory wardenContractAddress = $.wardenContractAddress;
+    if (!Strings.equal(wardenContractAddress, sourceAddress)) revert InvalidSourceChain();
+
+    IAxelarGateway gateway = IAxelarGateway($.axelarGateway);
+    if (!gateway.validateContractCall(commandId, sourceChain, sourceAddress, keccak256(payload)))
```

Comment:
Wouldn't it be better to `gateway.validateContractCall(commandId, $.wardenChain, $.wardenContractAddress, keccak256(payload))`? Seems like you can skip all the previous checks if done like this

Reply:
left as is for now